### ISSUE TITLE
Deep feature extraction and deployment

### ISSUE THREAD

**Issue by jrabary** [amplab/SparkNet#48]
Hi there,
We're trying to use caffe with spark for some video analytic task. Caffe is mainly used as feature extractor. Currently, our implementation is based on caffe javacpp preset. We have a lot of memory issue with this approach because of our way to manage the caffe network allocation and initialization. We'd like to test SparkNet as feature extractor but it seems that SparkNet doesn't have this feature yet. As far as I know it is just use for training and testing a neural network. Do you plan to use also SparkNet to deploy
a caffe network on spark and how can we add just a forward method on Net class ?

**Comment by robertnishihara** [comment]
Hi! That's a really good point. We added "forward" to Net.scala and wrote `src/main/scala/apps/FeaturizerApp.scala` to give one example of how to use it for featurizing data, see #49. 

At the moment, `CaffeNet.getData()` returns all of the data blobs from the network, which is inefficient. It would be simple to change this method to take the name of a blob and return that specific blob. If you plan on implementing that change or other changes, feel free to submit a pull request. This method can be optimized in other ways (e.g., writing into a buffer provided by the caller instead of allocating).

We are in the process of trying out JavaCPP.

**Comment by mazene** [comment]
Hi,
I am currently trying to use SparkNet  with distributed CPU instead of GPU (for some comparisons), I have set  Caffe with CPU_only config in caffe makefile configuration and then compiled SparkNet, then I submitted the CIFAR app locally : 
spark-submit --master local[*] --class apps.CifarApp  SparkNet/target/scala-2.10/sparknet-assembly-0.1-SNAPSHOT.jar 1
I am getting the following : F0108 11:18:43.727069 13912 split_layer.cpp:53] Cannot use GPU in CPU-only Caffe: check mode.

- What other modifications need to be done so that SparkNet perform the training on the CPU smoothly.

- what happens to the executor JVM thread (that originally owns the RDD partition data ) when control is given to the GPU for processing? what if another thread is scheduled on that executor while a previous thread is doing  computations on the GPU?

- how does  your design integrate with SPARK with the original spark DAG and task scheduler? is the cluster manager aware of the existence of GPU? what if the cluster manager schedule the task (JVM on a non-GPU enabled nodes)

Thank you.
